File: 2020-05-month-long-challenge/day14.py
#!/usr/bin/env python3

# Day 14: Implement Trie (Prefix Tree)
#
# Implement a trie with insert, search, and startsWith methods.
#
# Notes:
# - You may assume that all inputs are consist of lowercase letters a-z.
# - All inputs are guaranteed to be non-empty strings.

class Trie:
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = {} # basic types for the win
        

    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        """
        node = self.root
        for letter in word:
            if letter not in node:
                node[letter] = {}
            node = node[letter]
        node['#'] = True
        

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the trie.
        """
        node = self.root
        for letter in word:
            if letter not in node:
                return False
            node = node[letter]
        return '#' in node
        

    def startsWith(self, prefix: str) -> bool:
        """
        Returns if there is any word in the trie that starts with the given prefix.
        """
        node = self.root
        for letter in prefix:
            if letter not in node:
                return False
            node = node[letter]
        return True


# Tests
trie = Trie()
trie.insert("apple")
assert trie.search("apple") == True
assert trie.search("app") == False
assert trie.startsWith("app") == True
trie.insert("app");   
assert trie.search("app") == True

File: 2020-05-month-long-challenge/day14.js
#!/usr/bin/env nodejs

/*
 * Day 14: Implement Trie (Prefix Tree)
 *
 * Implement a trie with insert, search, and startsWith methods.
 *
 * Notes:
 * - You may assume that all inputs are consist of lowercase letters a-z.
 * - All inputs are guaranteed to be non-empty strings.
 */

/**
 * Initialize your data structure here.
 */
var Trie = function() {
    this.root = {};
};

/**
 * Inserts a word into the trie. 
 * @param {string} word
 * @return {void}
 */
Trie.prototype.insert = function(word) {
    node = this.root;
    for (letter of word) {
        if (node[letter] == undefined) {
            node[letter] = {};
        }
        node = node[letter];
    }
    node['#'] = true;
    
};

/**
 * Returns if the word is in the trie. 
 * @param {string} word
 * @return {boolean}
 */
Trie.prototype.search = function(word) {
    node = this.root;
    for (letter of word) {
        if (node[letter] == undefined) {
            return false;
        }
        node = node[letter];
    }
    return node['#'] != undefined;
};

/**
 * Returns if there is any word in the trie that starts with the given prefix. 
 * @param {string} prefix
 * @return {boolean}
 */
Trie.prototype.startsWith = function(prefix) {
    node = this.root;
    for (letter of prefix) {
        if (node[letter] == undefined) {
            return false;
        }
        node = node[letter];
    }
    return true;
};

// Tests
var trie = new Trie();
trie.insert("apple");
console.assert(trie.search("apple") == true);
console.assert(trie.search("app") == false);
console.assert(trie.startsWith("app") == true);
trie.insert("app");   
console.assert(trie.search("app") == true);
